# Basic Setup: Libraries + files

In [30]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [31]:
# Creating a path
path = r'C:\Users\zwier\anaconda_projects\15-01-2025 Instacart Basket Analysis'

In [32]:
# importing df_prods with os.path.join
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [33]:
# importing df_prods with os.path.join
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency Checks

In [35]:
# descriptive analysis
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Mixed-Type Data# Mixed-Type Data

# Create a dataframe

In [37]:
# Create a dataframe
df_test = pd.DataFrame()

In [38]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [52]:
# Checking results
df_test

mix
0     a
1     b
2     1
3  True

In [60]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\1522843687.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [62]:
# Changing data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [64]:
# Checking results
df_test

mix
0     a
1     b
2     1
3  True

# Check for mixed types

In [66]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\1522843687.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


# Missing Values

# Finding Missing Values

In [88]:
# Finding Missing Values in prod
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [90]:
# Finding Missing Values in orders
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [95]:
# subset of the dataframe containing only the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [97]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [103]:
# 1. Create a new variable that acts like a flag based on the missing value.
# 2. Impute the value with the mean or median of the column (if the variable is numeric).
# 3. Remove or filter out the missing data.

In [117]:
# Info
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [119]:
# Value Count
df_prods['product_name'].value_counts(dropna = False)

product_name
NaN                                                     16
Black House Coffee Roasty Stout Beer                     2
Adore Forever Body Wash                                  2
Gluten Free Organic Peanut Butter & Chocolate Cereal     2
Fiber 4g Gummy Dietary Supplement                        2
                                                        ..
Kosher Dill Mini Pickles                                 1
Chili Rice Scooter                                       1
All Natural Chicken Apple Sausage                        1
Quartered Artichoke Hearts                               1
Fresh Foaming Cleanser                                   1
Name: count, Length: 49673, dtype: int64

In [121]:
# desctryptive check
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Replacing missings

In [125]:
# code to replace missings with mean value 
# df['column with missings'].fillna(mean value, inplace=True)

In [127]:
# code to replace missings with median value
# df['column with missings'].fillna(median value, inplace=True)

In [129]:
# checking 
df_prods.shape

(49693, 5)

# clean new dataframe

In [131]:
# Lets create a clean new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [145]:
# checking 
df_prods_clean.shape

(49677, 5)

In [147]:
# Info
df_prods_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49677 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49677 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49677 non-null  int64  
 3   department_id  49677 non-null  int64  
 4   prices         49677 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 2.3+ MB


# deleting missings

In [151]:
# quick way to delete Nan and overwriting oryginal file - way 1
# df_prods.dropna(inplace = True)

In [149]:
# quick way to delete Nan and overwriting oryginal file - way 2
# df_prods.dropna(subset = [‘product_name’], inplace = True)

# Duplicates

# Finding Duplicates

In [155]:
# The following command will look for full duplicates within your dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [160]:
# Results
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [ ]:
# pandas has a handy function for just this purpose - df.drop_duplicates()

In [164]:
# checking shape
df_prods_clean.shape

(49677, 5)

In [166]:
# lets delete the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [176]:
# checking shape
df_prods_clean_no_dups.shape

(49672, 5)

In [178]:
# Double checking - new cleaned df
df_prods_clean_no_dups[df_prods_clean_no_dups.duplicated()]

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [182]:
# Exporting clean and checked file to prepered folder
# df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Exercise 4.5

# Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [188]:
# Checking Descryptives
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [ ]:
# 1. "order_number" - Check if the high values (e.g., 100) are consistent with user activity or result from data duplication or test accounts.
# 2. "days_since_prior_order" - Explore the distribution of 0 values to determine if these represent valid consecutive orders or potential data entry issues.
# 3. "Unnamed: 0" - If this is redundant, it can be removed to declutter the dataframe

# Check for mixed-type data in your df_ords dataframe.

In [192]:
# Check for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\2061706713.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\2061706713.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\2061706713.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\2061706713.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\zwier\AppData\Local

In [208]:
# Check for mixed types. Fixed (i hope)

for col in df_ords.columns:
    # Check if the column contains more than one unique data type
    unique_types = df_ords[col].apply(type).nunique()
    if unique_types > 1:
        print(f"Column {col} contains mixed types.")
    else:
        print(f"Column {col} is clean.")


Column Unnamed: 0 is clean.
Column order_id is clean.
Column user_id is clean.
Column eval_set is clean.
Column order_number is clean.
Column orders_day_of_week is clean.
Column order_hour_of_day is clean.
Column days_since_prior_order is clean.


# If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

## nothing to fix

Column Unnamed: 0 is clean.
Column order_id is clean.
Column user_id is clean.
Column eval_set is clean.
Column order_number is clean.
Column orders_day_of_week is clean.
Column order_hour_of_day is clean.
Column days_since_prior_order is clean.

# Run a check for missing values in your df_ords dataframe.

In [215]:
# Finding Missing Values in orders
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

days_since_prior_order: Contains 206,209 missing values.
Missing values likely represent the first order for each user in the dataset. Since there is no prior order to calculate the gap, this column would naturally have a NaN for such rows.



# Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice

In [223]:
# Fill missing values with 0
df_ords['days_since_prior_order'].fillna(0, inplace=True)

C:\Users\zwier\AppData\Local\Temp\ipykernel_12972\4265045068.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(0, inplace=True)


Filling missing values with 0 to indicate that it is the user's first order. This approach is logical because a "0-day gap" reflects the absence of prior orders. No clue why this failed. 

# Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [229]:
# Checking for duplicates
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [233]:
# Counting Duplicate Rows
duplicate_count = df_ords.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_count}")


Number of duplicate rows: 0


# Answer: There are non

# Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

# Answer: There are non

In [241]:
# Exporting clean and checked file to prepered folder
# df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))